pkg/policy/prefilter.go

```diff
@@ -0,0 +1,303 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package policy
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"os/exec"
+	"path"
+	"sync"
+	"syscall"
+
+	"github.com/cilium/cilium/pkg/bpf"
+	"github.com/cilium/cilium/pkg/maps/cidrmap"
+)
+
+type preFilterMapType int
+
+const (
+	prefixesV4Dyn preFilterMapType = iota
+	prefixesV4Fix
+	prefixesV6Dyn
+	prefixesV6Fix
+	mapCount
+)
+
+const (
+	maxLKeys = 1024 * 64
+	maxHKeys = 1024 * 1024 * 20
+)
+
+type preFilterMaps [mapCount]*cidrmap.CIDRMap
+
+type preFilterConfig struct {
+	dyn4Enabled bool
+	dyn6Enabled bool
+	fix4Enabled bool
+	fix6Enabled bool
+}
+
+// PreFilter holds global info on related CIDR maps participating in prefilter
+type PreFilter struct {
+	Maps     preFilterMaps
+	Config   preFilterConfig
+	Revision int64
+	Mutex    sync.RWMutex
+}
+
+// WriteConfig dumps the configuration for the corresponding header file
+func (p *PreFilter) WriteConfig(fw io.Writer) {
+	p.Mutex.Lock()
+	defer p.Mutex.Unlock()
+
+	fmt.Fprintf(fw, "#define CIDR4_HMAP_ELEMS %d\n", maxHKeys)
+	fmt.Fprintf(fw, "#define CIDR4_LMAP_ELEMS %d\n", maxLKeys)
+
+	fmt.Fprintf(fw, "#define CIDR4_HMAP_NAME %s\n", path.Base(p.Maps[prefixesV4Fix].String()))
+	fmt.Fprintf(fw, "#define CIDR4_LMAP_NAME %s\n", path.Base(p.Maps[prefixesV4Dyn].String()))
+	fmt.Fprintf(fw, "#define CIDR6_HMAP_NAME %s\n", path.Base(p.Maps[prefixesV6Fix].String()))
+	fmt.Fprintf(fw, "#define CIDR6_LMAP_NAME %s\n", path.Base(p.Maps[prefixesV6Dyn].String()))
+
+	if p.Config.fix4Enabled {
+		fmt.Fprintf(fw, "#define CIDR4_FILTER\n")
+		if p.Config.dyn4Enabled {
+			fmt.Fprintf(fw, "#define CIDR4_LPM_PREFILTER\n")
+		}
+	}
+	if p.Config.fix6Enabled {
+		fmt.Fprintf(fw, "#define CIDR6_FILTER\n")
+		if p.Config.dyn6Enabled {
+			fmt.Fprintf(fw, "#define CIDR6_LPM_PREFILTER\n")
+		}
+	}
+}
+
+func (p *PreFilter) dumpOneMap(which preFilterMapType, to []string) []string {
+	if p.Maps[which] == nil {
+		return to
+	}
+	return p.Maps[which].CIDRDump(to)
+}
+
+// Dump dumps revision and CIDRs as string slice of all participating maps
+func (p *PreFilter) Dump(to []string) ([]string, int64) {
+	p.Mutex.Lock()
+	defer p.Mutex.Unlock()
+	for i := prefixesV4Dyn; i < mapCount; i++ {
+		to = p.dumpOneMap(i, to)
+	}
+	return to, p.Revision
+}
+
+func (p *PreFilter) selectMap(ones, bits int) preFilterMapType {
+	if bits == 32 {
+		if ones == bits {
+			return prefixesV4Fix
+		}
+		return prefixesV4Dyn
+	} else if bits == 128 {
+		if ones == bits {
+			return prefixesV6Fix
+		}
+		return prefixesV6Dyn
+	} else {
+		return mapCount
+	}
+}
+
+// Insert inserts slice of CIDRs (doh!) for the latest revision
+func (p *PreFilter) Insert(revision int64, cidrs []net.IPNet) error {
+	var undoQueue []net.IPNet
+	var ret error
+
+	p.Mutex.Lock()
+	defer p.Mutex.Unlock()
+	if p.Revision != revision {
```

Comment:
Revision should be optional here and for delete. The user can opt-in to use this transactional feature if desired.

Reply:
Okay, will make it optional.

Reply:
Fixed.